### PR TITLE
Sidekiq Pro & SuperFetch support

### PR DESCRIPTION
**Pull request by holstvoogd**
I think this covers all of what I have added/changed so far:

SuperFetch support:
- Defer fetching jobs to BasicFetch, or SuperFetch if configured.
- Patches BasicFetch for compatability with SuperFetch and Throttled.
- Adds Sidekiq::Throttled.recover! for 'unthrottling' jobs that have been recovered by SuperFetch.

Enhanced queues:
- Disable queue pausing & enhanced tabs when using Sidekiq::Pro.

Sidekiq::Pro testing:
- Added Appriasals for sk-pro 5.3 & 5.5; the latest 2 minor versions. Local preinstall required 
- Spec'd with both Basic- & SuperFetcher

Changed default behaviour:
- Strict default is now taken from the Sidekiq::DEFAULTS & thus defaults to true in 6.1 and later

Documentation:
- Basic setup & enhanced queues default behavior
- Job recovery

TODO (maybe?):
- Might still need to add something for checking the queue 'pausing' when throttled works
- Same for the enhanced queues disabled mode
- Testing instructions for pro

### CODE REVIEW

**Comment by holstvoogd** [comment]
@ixti Took a bit due to the flu, but I think I'm getting there :)

File: spec/sidekiq_helper.rb [hunk 1]
```diff
@@ -17,6 +17,11 @@ def jid
 Sidekiq.configure_server(&configure_redis)
 Sidekiq.configure_client(&configure_redis)
 
+# Makes sidekiq(-pro) more testable
+# rubocop: disable Style/GlobalVars
+$TESTING = true
+# rubocop: enable Style/GlobalVars
+
```
Comment (ixti) [inline spec/sidekiq_helper.rb]: I think we can move this to spec_helper.rb and same-line disable too:

``` ruby
# Makes sidekiq(-pro) more testable
$TESTING = true # rubocop: disable Style/GlobalVars
```

File: README.md [hunk 2]
```diff
@@ -238,6 +238,44 @@ Rails.application.routes.draw do
...
+this gem to clear throttles for jobs it recovers. You can do this by providing
+a block to `Sidekiq.super_fetch!` during setup:
+
+``` ruby
+require "sidekiq/throttled"
+
+# Enable SuperFetch & configure it to notify Throttled of recovered jobs
+Sidekiq.super_fetch! do |msg|
+  Sidekiq::Throttled.recover!(msg)
+end
```
Comment (ixti) [inline README.md]: Should we do this automatically as part of the setup?

**Review by ixti — COMMENTED** [review]
WDYT about doing a bigger refactoring and:

* Remove queue pauser
* Remove communicator
* Drop Sidekiq < 6.4 support

I can do that as a separate PR to avoid scope creep. But that will make the changeset you need in this PR smaller.

File: README.md (hunk 2 above)
Comment (holstvoogd) [reply README.md]: I am not sure. I thought that might be a little to 'smart' and lead to suprises or unwanted things like overwriting the fetcher if you have a custom one. And what to do when you need your own code in there too? (For instance, we use sideiq-unique-jobs too & that needs to clean up locks from this block.)

On the other hand, the most common use case would be: If I have sidekiq pro, I want to use super_fetch! _and_ make sure jobs are recovered.

We could do something like this in the setup so any config made before the Throttled setup are retained:
```ruby
original_handler =  Sidekiq[:fetch].orphan_handler
Sidekiq.super_fetch! do |msg|
  Sidekiq::Throttled.recover!(msg)
  original_handler.call(msg)
end
```

Or we can just say, if you want to do anything other then use throttled + super_fetch, you have to pass it explicitly and make sure you setup orphan handling for it. If you don't provide a fetcher, we construct one with the orphan handling.

What do you think?

**Comment by holstvoogd** [comment]
@ixti  A bigger refactor would make the entire code a lot cleaner too. I think that is a good idea. Would you still want a seperate gem for pausable queues?

There would be some overlap, both would require the same BasicFetch patche(s) to make it ingore queues, but I think that is fine.

**Comment by holstvoogd** [comment]
I've got some suprise time to work on this today while at the office, so I'll get started on cleaning up the queues etc & make some additional PRs to split it all up 👍

**Comment by holstvoogd** [comment]
I've remove the queue pauser & pre 6.4 support and then rebased on top of that. I think its a good idea, saves some complexity with all the `if`s.

I haven't looked into the orphan_handler any further yet, but I it is probably best to go with a setup where we follow the same pattern as with the fetcher: store the old handler, build a new one that calls it _and_ recovers the throttled jobs.

**Comment by ixti** [comment]
> @ixti A bigger refactor would make the entire code a lot cleaner too. I think that is a good idea. Would you still want a seperate gem for pausable queues?
> There would be some overlap, both would require the same BasicFetch patche(s) to make it ingore queues, but I think that is fine.


Thanks for your clean-up PRs. Just merged them. I want to have queues pausing as a separate gem, yes. Thus, at some point, I think we will need to come up with some sort of composable/injectable API. Another reason I want this so bad, is I want sidekiq-throttled to play nice with [sidekiq-ultimate](https://github.com/sensortower/sidekiq-ultimate/blob/master/lib/sidekiq/ultimate/fetch.rb) :D

File: README.md (hunk 2 above)
Comment (ixti) [reply README.md]: I'm not a Sidekiq-Pro user, so I don't have strong feelings one way or the other. I share your sentiments that automatic setup might be surprising. Thus, I leave the final decision for you to make.

**Comment by holstvoogd** [comment]
👍 I've updated this PR. I'll still have a look at the config/setup, I'll see if I can do that later today still.

As for an api, I'm thinking wether a middleware chain could work if only implemented for retrieve_work.
I imagine that a middleware api where each layer would further manipulate the queues to fetch from could work.

**Comment by holstvoogd** [comment]
After some discussion with colleagues, automatically setting up a wrapper for the orphan handling seems safest. So I've added that to the setup & update the readme etc. I think this should probably be good now.

(And at some point I hope to learn to run rubocop before pushing ;))

**Comment by ixti** [comment]
> As for an api, I'm thinking wether a middleware chain could work if only implemented for retrieve_work. I imagine that a middleware api where each layer would further manipulate the queues to fetch from could work.

I was thinking about somewhat like that, yes. For now the plan is this, we can release 1.0.0.pre1, and as soon as queue pluggable pause will be ready (and we prove it plays nice with sidekiq-throttled) we will release 1.0.0.

**Comment by ixti** [comment]
I'll review the changes later this week.

File: lib/sidekiq/throttled.rb [hunk 3]
```diff
@@ -94,8 +94,10 @@ def throttled?(message)
       # @return [Void]
       def recover!(message)
         with_strategy_and_job(message) do |strategy, jid, args|
-          strategy.finalize! jid, *args
+          strategy.finalize!(jid, *args)
         end
+      rescue
+        false
```
Comment (ixti) [inline lib/sidekiq/throttled.rb:100]: It's worth to show the exception in logs - will save lots of headache in case of weird situations ;))

Comment (holstvoogd) [reply lib/sidekiq/throttled.rb:100]: hehe, fair enough :) I'll add some to the throttled? method too I guess. for when parse errors occur there that is.

**Comment by ixti** [comment]
Yes, please.

File: lib/sidekiq/throttled.rb (hunk 3 above)
Comment (ixti) [reply lib/sidekiq/throttled.rb:100]: Yes, please! :D

File: README.md [hunk 4]
```diff
@@ -235,6 +245,10 @@ This library aims to support work with following [Sidekiq][sidekiq] versions:
 * Sidekiq 6.4.x
 * Sidekiq 6.5.x
 
+And the following Sidekiq Pro versions:
+
+* Sidekiq Pro 5.3.x
+* Sidekiq Pro 5.5.x
```
Comment (ixti) [inline README.md:251]: I think we should not list Sidekiq PRO as "officially supported". More like "it should work", unless you're willing (and I'm happy to welcome you) to step up as Sidekiq PRO compatibility maintainer :D

**Comment by holstvoogd** [comment]
Haven't found time to update the PR yet, but will eventually i hope 🤞 
Agreed on the supported bit, I'll be happy to keep it compatible etc, but as long we can't incorporate the pro tests in CI; it does seem smart to manage expectations haha

We have been running this in production for a while now and it works well with regards to the reliable fetch part, so that is good news ;) 
Something else I'm looking into: we run into an issue where we are constantly fetching & requeueing jobs on a queue while other queues are full of runnable jobs. So I think I broke the little pause of queues when a throttled job is found 🤔 Anyway, more to come on this

**Comment by ckhall** [comment]
Hi all.  Just curious about the state of this PR. Thanks!

**Comment by holstvoogd** [comment]
I kind of forgot about it tbh 😅 I'll see if I find time to rebase it, but I'm a bit short on free time atm so i can't really promise anything.

fwiw, we've been running this in production since august, haven't seen any issues.

**Comment by ckhall** [comment]
@holstvoogd that's for the quick response.  Thanks for the update.  I was curious about this in your last comment

> Something else I'm looking into: we run into an issue where we are constantly fetching & requeueing jobs on a queue while other queues are full of runnable jobs. So I think I broke the little pause of queues when a throttled job is found 🤔 Anyway, more to come on this

Were you able to sort this issue out?

**Comment by holstvoogd** [comment]
I forgot about that. We haven't had many issues with jobs not being pulled since then afaik. 

Although last week we ran into an issues when we had ~20 batches of 25k jobs, throttled 1 per batch & queued the entire batch at a time. So the workers would start a job & spend a lot of time moving the rest of the batch to the back of the queue. This would cause other jobs on the queue to be delayed by hours.

Beter queue management fixed that, i.e. move the bulk jobs to a seperate queue. But it still costs a lot of time to move all the jobs.. I'm thinking if it would be possible to be smarter about this, or if it is just innevitable given the strategy for fetching. Maybe it would be possible to move some of the logic into redis & have it only fetch non-throttled jobs without the requeueing.

Anyway, I think both issues are the same issue really: to many sequenctial jobs within the same throttle. Better would be scheduling them over some period or making sure they are atleast interleaved if there are multiple batches.
Also, don't put bulk processing jobs on the same queue as email ;)

**Comment by ixti** [comment]
I've got back accees to the original repo - so all further development will be done there. Already cloned super-fetch and sidekiq-pro-support branches in there and will work on merging them after v1.0.0 release (that will bring sidekiq-7.0 support).

**Pull request closed without merging**